{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# circRNA snakemake pipeline for Biodonostia"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Authors: Leire Iparragirre, Inigo Prada and David Otaegi\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I am going to develop a small snakemake pipeline for David Otaegui and Leire Iparraguirre at Biodonostia for analyzing circRNA sequencing data. The pipeline will have the following steps:\n",
    "\n",
    "# Sequencing QC and Sequence Trimming\n",
    "\n",
    "Nowadays, Illumina sequencing quality is good enough to skip this step. And it is not adviced any more to perform agresive sequence trimming on RNA-seq. Hence, I will include a simple sequencing quality checking step (fastqc-0.11.7) and lets trim afterwards If required ( I hope it is not :-)\n",
    "\n",
    "# STAR index creation\n",
    "\n",
    "   #### Overall explanation of this step\n",
    "\n",
    "The pipeline will include a step were it is checked if STAR has its index of the genome and created if necessary. David and Leire, don't bother about this. An index is an efficient data structure to speed-up the alignment speed and reduce memory usage. If you are still interested on this (amazing) data structures, it is perfectly explained in this blog post:\n",
    "\n",
    "http://blog.thegrandlocus.com/2016/07/a-tutorial-on-burrows-wheeler-indexing-methods\n",
    "\n",
    "The reference data has been downloaded from Illumina iGenomes the 20-07-18 from this link https://support.illumina.com/sequencing/sequencing_software/igenome.html \n",
    "\n",
    "The Illumina iGenomes page includes a changelog, so If you wish to repeat the analysis be sure that you are working with the same files :)\n",
    "\n",
    "ftp://ussd-ftp.illumina.com/CHANGE_LOG.txt\n",
    "\n",
    "#### Technical explanation\n",
    "\n",
    " In this step we need to build a efficient data structure based on suffix arrays so that STAR can align million of reads per minute as efficient as possible. This is done with the following command:\n",
    " \n",
    " STAR --runThreadN {threads} --runMode genomeGenerate --outStd Log {log} --genomeDir {input.ref_genome_dir} --genomeFastaFiles {input.ref_genome}\n",
    " \n",
    " Where:\n",
    "     - --runThreadN {threads} {Threads}: Number of threads to use to build the index (to increase speed)\n",
    "     - --runMode genomeGenerate: Tell STAR to build the suffix array of the reference genome\n",
    "     - --outStd Log {log}: Where to save log files\n",
    "     - --genomeDir {input.ref_genome_dir}: Where to save the STAR index (This directory will contain) a bunch of files required  by STAR for aligning the read\n",
    "     --genomeFastaFiles {input.ref_genome} Where is the Genome fasta file\n",
    "     \n",
    "     \n",
    "\n",
    "# Alignment \n",
    "\n",
    "#### Overall explanation of this step\n",
    "\n",
    "Once we have generate STAR index, we are ready to align the reads to reference genome. The general idea of this step is to determine to location of every read on the genome. To do so, we align the reads to the efficient representation of the genome we have created before. I completely skip all the alignment details, if you are interested on alignment algorithms, you can read about the alignment algorithms on chapter 2 (sections 2.1-2.5) of the following book:\n",
    "\n",
    "Durbin, R., Eddy, S.R., Krogh, A. and Mitchison, G., 1998. Biological sequence analysis: probabilistic models of proteins and nucleic acids. Cambridge university press.\n",
    "\n",
    "Once aligned, STAR outputs a bunch of files, been the most important one the bam file. The bam (binary alignment/map format ) is a file that contains crazy amount of information about how every read has align to the genome. Each read will contain encoded information as: alignment quality, is the alignment spliced?, are there any mismatches in the alignment? and a large etc.\n",
    "\n",
    "You can read more about the sam format on the following links:\n",
    "\n",
    "lightweight link:\n",
    "\n",
    "https://genome.sph.umich.edu/wiki/SAM\n",
    "\n",
    "heavyweight link:\n",
    "\n",
    "https://samtools.github.io/hts-specs/SAMv1.pdf\n",
    "\n",
    "\n",
    "#### Technical explanation \n",
    "\n",
    "In this step we align the reads to the genome with the following command\n",
    "\n",
    "\n",
    "STAR --chimSegmentMin 10 --runThreadN {threads} --outSAMtype {params.bam_output} --genomeDir {params.reference} --readFilesCommand zcat --outFileNamePrefix {params.substr} --outStd Log {log} --readFilesIn {input.r1} {input.r2}\n",
    "\n",
    " - --chimSegementMin: Minimum length of chimeric reads (the ones indicating circRNA)\n",
    " - --runThreadN: Number of threads to use (speed up)\n",
    " - --genomeDir: Where is the index of the reference genome\n",
    " - --outFileNamePrefix: Output file name\n",
    " - --outStd Log: Where to put the log files\n",
    " - --readFilesCommand zcat: Assumes compressed input (tar.gz) \n",
    " --readFilesIn: What are the input files\n",
    "\n",
    "# Aligment QC\n",
    "\n",
    "After the alignment one of the things that most scare biologist and bioinformaticians is the fact that reads align poorly or do not even align. To check so, we use the program samtools (v1.8 using htslib 1.8) which has severall tools to manage sam/bam files.\n",
    "\n",
    "#### Technical explanation \n",
    "\n",
    "Command:\n",
    "\n",
    "\n",
    "samtools flagstat input.bam\n",
    "\n",
    "In this case, we will use samtools flagstat, which computes summary statistics of the alignment. The resulting samtools flagstat looks as the following:\n",
    "\n",
    "20968800 + 0 in total (QC-passed reads + QC-failed reads)\n",
    "0 + 0 duplicates\n",
    "20968800 + 0 mapped (100.00%:nan%)\n",
    "20968800 + 0 paired in sequencing\n",
    "11431237 + 0 read1\n",
    "9537563 + 0 read2\n",
    "71098 + 0 properly paired (0.34%:nan%)\n",
    "6633306 + 0 with itself and mate mapped\n",
    "14335494 + 0 singletons (68.37%:nan%)\n",
    "0 + 0 with mate mapped to a different chr\n",
    "0 + 0 with mate mapped to a different chr (mapQ>=5)\n",
    "\n",
    "\n",
    "Where:\n",
    "\n",
    "    -Mapped: read considered as aligned\n",
    "    - Paired in sequencing: If the reads were paired (paired-end sequencing)\n",
    "    - read1: From the total (including unmmaped )reads, how many of them are the read 1 (forward read in paired end sequencing)\n",
    "    - read2: From the total (including unmapped) reads, how many of them are the read 2  (reverse read )\n",
    "    - properly paired: How many reads have align in the proper orientation (reverse and forward reads pointing at each other)\n",
    "    - with itself and mate mapped: Both reads  (forward and reverse) mapped\n",
    "    - Singletons: One of the paired reads is mapped and the other one is unmapped\n",
    "    -with mate mapped to a different chr (mapQ>=5): Paired read were one of the reads has align to one chromosome and the other was is aligned to a different chromosome (this reads indicate, for example, interchromosomal translocations)\n",
    "    -mapq>=5: Probability that the read is mapped wrongly aprox 0.31\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "# CircRNA detection\n",
    "\n",
    "\n",
    "David and Leire, you are familiar with this step :)\n",
    "\n",
    "The overall idea for this step is to detect those reads that overlap the backspliced junction of the circRNA.To the detect the circRNA we will use CIRCexplorer2 (v2.3.3).First, Circexplorer parses the chimeric junctions (which indicate circRNA ) from the STAR generated bam files, and, then, the circexplorer parsed file is annotated with the circRNA\n",
    "\n",
    "\n",
    "\n",
    "#### Technical explanation \n",
    "\n",
    "Parse star output:\n",
    "\n",
    "CIRCexplorer2 parse -t {params.aligner} -b {output.parsed_junctions} {input.junctions}\n",
    "\n",
    "where:\n",
    "\n",
    "-t: Which aligner I have use\n",
    "-b: File name output\n",
    "{input.junctions}: STAR backspliced junction output\n",
    "\n",
    "Annotate STAR output:\n",
    "\n",
    "CIRCexplorer2 annotate -r {input.ref_gene} -g {input.hg38} -b {input.parsed_junctions} -o {output.circrna} \n",
    "\n",
    "where:\n",
    "\n",
    "\n",
    "-r: Gene annotation file of the reference genome\n",
    "-g: REference genome fasta file\n",
    "-b: Circexplorer parsed junctions\n",
    "-o output file\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
